README.md
# BuildVRFIndex.py #
----------

## About ##
**BuildVRFIndex.py** is a Python application created from the buildIndex()
function in VRF Search Tool v0.0.10-beta.  I created this stand-alone
application so that an index of VRF names and their locations could be
updated without having to run an application containing this function.
This may be useful if run as a scheduled job.

## Dependencies ##
Exscript module [https://github.com/knipknap/exscript/](https://github.com/knipknap/exscript/)

## Requirements ##
1. This application is hard-coded to use the SSH2 protocol; If SSH v2 is not
   enabled on your router(s), you will need to:
   * Add `ip ssh version 2` to your Cisco router(s) configuration and any 
   associated access-list changes.
   or
   * Alter the `default_protocol` variable in the `get_hosts_from_file` function
   to use a different protocol enabled on your router(s).
2. A valid username/password.
3. Presence of a `routers.txt` file containing a list, one per line, of 
   Hostnames or IP Addresses of routers on which VRF tunnels are configured.
   (If one is not provided, the application will create an example file for
   the user.)
4. Presence of a `settings.cfg` file containing optional configuration settings
   to fully automate or alter the functionality of the application. (If one is 
   not provided, the application will create an example file for the user.)
   Absence of custom settings in the `settings.cfg` file will cause the application
   to run in interactive mode with default settings.
5. If a password is stored in `settings.cfg` it must be base64-encoded.

## Assumptions ##
1. This application was written for use on Cisco IOS devices and cannot be
   guaranteed to work on other makes/model routers.
2. This application assumes that you have enable privileges on each router
   in order to execute the `show running-config` command.  If you do not
   have sufficient user privileges, this application will not work as
   designed.

## Limitations ##
1. This application uses the same username/password to access ALL routers. If
   your routers use unique usernames/passwords, then this script will not work.

## Functionality ##
![](BuildVRFIndex.png)

1. Upon execution, the application will search its parent directory for the 
   presence of a file named `settings.cfg`.
   * If this file exists, the application will proceed to Step 3.
   * If this file does not exist, the application will continue to Step 2.
2. The application will create an example `settings.cfg` file in the parent
   directory and continue to Step 3.
3. The application will open the `settings.cfg` file and retrieve any custom
   settings from the file and continue to Step 4.
4. The application will search its parent directory for the presence of a file
   named `routers.txt`.
   * If this file exists, the application will continue to Step 5.
   * If this file does not exist, the application will proceed to Step 6.
5. The application will check to see if a username and password was specified
   in the `settings.cfg` file.
   * If a username and password was set, the application will proceed to Step 8.
   * If a username and password was not set, the application will continue to Step 7.
6. The application will display instructions for the use of the `routers.txt`
   file.  The application will also create an example file and place it in 
   its parent directory end exit (END).
7. The application will ask the user for their username and password.  These
   credentials will be used to login to each of the routers listed in the 
   `routers.txt` file.  Continue to Step 8.
8. The application will log in to each of the routers listed in the
   `routers.txt` file using the username and password provided by the user or
   obtained from the `settings.cfg` file. 
   Using the `show run | include crypto keyring` command, the application 
   will collect all the VRFs configured on each router and write them to 
   the `index.txt.tmp` file.  Continue to Step 9.
9. Removing all the unnecessary information from the `index.txt.tmp` file,
   the application will copy the VRF Name, Customer Peer IP and Local Peer IP
   (separated by commas) into an `index.txt` file.  The application will then
   remove the temporary file, `index.txt.tmp` and exit (END).
   Note: If the `index.txt` file already exists, it will be overwritten during 
   this step.


CHANGELOG.md
# BuildVRFIndex.py #
---

## BuildVRFIndex v0.0.9-alpha (2014-03-17) ##
* Replaced tab with four spaces.
* Replaced ' with " to be consistent throughout the file.
* Corrected problem where application would fail if logFileDirectory variable 
  in settings.cfg was blank.

## BuildVRFIndex v0.0.8-alpha (2013-09-09) ##
* Corrected makedirs() functionality: Directories with a trailing backslash
  in the config file were not being created thereby causing the application
  to fail.
* Moved logFileDirectory makedirs() function such that the
  directory would only be created if/when the parent function was called
  instead of creating the directory whenever the application executed.

## BuildVRFIndex v0.0.7-alpha (2013-08-29) ##
* Added basic logging to file to track results if application has to connect
  to a router to run buildIndex().
* Suppressed error SPAM from stdout by adding stderr=(open(os.devnull, 'w'))
  to the Queue() function. (Errors are still written to the log.)

## BuildVRFIndex v0.0.6-alpha (2013-08-28) ##
* Added functionality to specify configFile from the command line.
* Updated README.md

## BuildVRFIndex v0.0.5-alpha (2013-08-23) ##
* Removed Python modules not required by this application.
* Minor corrections to output spacing.

## BuildVRFIndex v0.0.4-alpha (2013-08-20) ##
* Added additional comments to code, configFile, routerFile.
* Added configFile functionality to give application the ability to retrieve
  user-specified settings from a config file.  Application use now extended
  such that the list of routers, index and respective paths can be specified
  in the file.  Application can also use configured username and password.

## BuildVRFIndex v0.0.3-alpha (2013-08-16) ##
* Changed the configFile format in such a way that a single configFile
  can be used for my different VRF applications.
* Updated error messages so they reflect the actual filename as read
  from the configFile.
* Rewrote "Building index..." message so it does not take up most of the
  screen when working with a large batch of routers.
* Updated example configFile output to reflect new changes in configFile format.

## BuildVRFIndex v0.0.2-alpha (2013-08-15) ##
* Added instructions for creating a password to settings.cfg
* Cleaned up module importing
* Removed quotations around variables in function that writes the example
  configFile (not necessary)

## BuildVRFIndex v0.0.1-alpha (2013-08-15) ##
* Alphabetized functions

## BuildVRFIndex v0.0.1-alpha (2013-08-14) ##
* Initial commit


BuildVRFIndex.py
#!/usr/bin/env python
#
# BuildVRFIndex.py
# Copyright (C) 2013-2014 Aaron Melton <aaron(at)aaronmelton(dot)com>
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.


import argparse     # Required to read arguments from the command line
import base64       # Required to decode password
import ConfigParser # Required for configuration file
import datetime     # Required for date format
import Exscript     # Required for SSH & queue functionality
import re           # Required for REGEX operations
import sys          # Required for printing without newline
import os           # Required to determine OS of host

from argparse                   import ArgumentParser, RawDescriptionHelpFormatter
from base64                     import b64decode
from ConfigParser               import ConfigParser
from datetime                   import datetime
from Exscript                   import Account, Queue, Host, Logger
from Exscript.protocols         import SSH2
from Exscript.util.file         import get_hosts_from_file
from Exscript.util.log          import log_to
from Exscript.util.decorator    import autologin
from Exscript.util.interact     import read_login
from Exscript.util.report       import status,summarize
from re                         import sub
from sys                        import stdout
from os                         import getcwd, makedirs, name, path, remove, system


class Application:
# This class was created to provide me with an easy way to update application
# details across all my applications.  Also used to display information when
# application is executed with "--help" argument.
    author = "Aaron Melton <aaron@aaronmelton.com>"
    date = "(2014-03-17)"
    description = "Creates a CSV of VRF Names across multiple Cisco routers"
    name = "BuildVRFIndex.py"
    url = "https://github.com/aaronmelton/BuildVRFIndex"
    version = "v0.0.9-alpha"


logger = Logger()   # Log stuff
@log_to(logger)     # Logging decorator; Must precede buildIndex!
@autologin()        # Exscript login decorator; Must precede buildIndex!
def buildIndex(job, host, socket):
# This function builds the index file by connecting to the router and extracting all
# matching sections.  I chose to search for "crypto keyring" because it is the only
# portion of a VPN config that contains the VRF name AND Peer IP.  Caveat is that
# the program temporarily captures the pre-shared key.  "crypto isakmp profile" was not
# a suitable query due to the possibility of multiple "match identity address" statements

    stdout.write(".")                   # Write period without trailing newline
    socket.execute("terminal length 0") # Disable user-prompt to page through config
                                        # Exscript doesn't always recognize Cisco IOS
                                        # for socket.autoinit() to work correctly

    # Send command to router to capture results
    socket.execute("show running-config | section crypto keyring")

    # Open indexFileTmp to temporarily hold router output
    with open(indexFileTmp, "a") as outputFile:
        try:
            outputFile.write(socket.response)   # Write contents of running config to output file
        
        # Exception: indexFileTmp could not be opened
        except IOError:
            print
            print "--> An error occurred opening "+indexFileTmp+"."

    socket.send("exit\r")   # Send the "exit" command to log out of router gracefully
    socket.close()          # Close SSH connection
    cleanIndex(indexFileTmp, host)  # Execute function to clean-up the index file
    
def cleanIndex(indexFileTmp, host):
# This function strips all the unnecessary information collected from the router leaving
# only the VRF name, remote Peer IP and local hostname or IP

    try:
        # If the temporary index file can be opened, proceed with clean-up
        with open(indexFileTmp, "r") as srcIndex:

            try:
                # If the actual index file can be opened, proceed with clean-up
                # Remove unnecessary details from the captured config
                with open(indexFile, "a") as dstIndex:
                    # Use REGEX to step through config and remove everything but
                    # the VRF Name, Peer IP & append router hostname/IP to the end
                    a = srcIndex.read()
                    b = sub(r"show running-config \| section crypto keyring.*", "", a)
                    c = sub(r"crypto keyring ", "", b)
                    d = sub(r".(\r?\n)..pre-shared-key.address.", ",", c)
                    e = sub(r".key.*\r", ","+host.get_name(), d)
                    f = sub(r".*#", "", e)
                    dstIndex.write(f)    # Write cleaned-up output to indexFile

            # Exception: actual index file could not be opened
            except IOError:
                print
                print "--> An error occurred opening "+indexFile+"."

    # Exception: temporary index file could not be opened
    except IOError:
        print
        print "--> An error occurred opening "+indexFileTmp+"."
    
    # Always remove the temporary index file
    finally:
        remove(indexFileTmp)    # Critical to remove temporary file as it contains passwords!

def fileExist(fileName):
# This function checks the parent directory for the presence of a file
# Returns true if found, false if not

    try:
        # If file can be opened, it must exist
        with open(fileName, "r") as openedFile:
            return True    # File found

    # Exception: file cannot be opened, must not exist
    except IOError:
        return False    # File NOT found
        
def routerLogin():
# This function prompts the user to provide their login credentials and logs into each
# of the routers before calling the buildIndex function to extract relevant portions of
# the router config.  As designed, this function actually has the capability to login to
# multiple routers simultaneously.  I chose to not allow it to multi-thread given possibility
# of undesirable results from multiple threads writing to the same index file simultaneously

    try:# Check for existence of routerFile; If exists, continue with program
        with open(routerFile, "r"): pass
        
        # Read hosts from specified file & remove duplicate entries, set protocol to SSH2
        hosts = get_hosts_from_file(routerFile,default_protocol="ssh2",remove_duplicates=True)

        if username == "":          # If username is blank
            print
            account = read_login()  # Prompt the user for login credentials

        elif password == "":        # If password is blank
            print
            account = read_login()  # Prompt the user for login credentials

        else:                       # Else use username/password from configFile
            account = Account(name=username, password=b64decode(password))
        
        # Minimal message from queue, 1 threads, redirect errors to null
        queue = Queue(verbose=0, max_threads=1, stderr=(open(os.devnull, "w")))
        queue.add_account(account)              # Use supplied user credentials
        print
        stdout.write("--> Building index...")   # Print without trailing newline
        queue.run(hosts, buildIndex)            # Create queue using provided hosts
        queue.shutdown()                        # End all running threads and close queue
        
        # If logFileDirectory does not exist, create it.
        if not path.exists(logFileDirectory): makedirs(logFileDirectory)

        # Define log filename based on date
        logFilename = logFileDirectory+"BuildVRFIndex_"+date+".log"

        # Check to see if logFilename currently exists.  If it does, append an
        # integer onto the end of the filename until logFilename no longer exists
        incrementLogFilename = 1
        while fileExist(logFilename):
            logFilename = logFileDirectory+"BuildVRFIndex_"+date+"_"+str(incrementLogFilename)+".log"
            incrementLogFilename = incrementLogFilename + 1

        # Write log results to logFile
        with open(logFilename, "w") as outputLogFile:
            try:
                outputLogFile.write(summarize(logger))

            # Exception: router file was not able to be opened
            except IOError:
                print
                print "--> An error occurred opening "+logFileDirectory+logFile+"."

    # Exception: router file could not be opened
    except IOError:
        print
        print "--> An error occurred opening "+routerFile+"."


# Check to determine if any arguments may have been presented at the command
# line and generate help message for "--help" switch
parser = ArgumentParser(
    formatter_class=RawDescriptionHelpFormatter,description=(
        Application.name+" "+Application.version+" "+Application.date+"\n"+
        "--\n"+
        "Description: "+Application.description+"\n\n"+
        "Author: "+Application.author+"\n"+
        "URL:    "+Application.url
    ))
# Add additional argument to handle any optional configFile passed to application
parser.add_argument("-c", "--config", dest="configFile", help="config file", default="settings.cfg", required=False)
args = parser.parse_args()      # Set "args" = input from command line
configFile = args.configFile    # Set configFile = config file from command line OR "settings.cfg"


# Determine OS in use and clear screen of previous output
if name == "nt":    system("cls")
else:               system("clear")

# PRINT PROGRAM BANNER
print Application.name+" "+Application.version+" "+Application.date
print "-"*(len(Application.name+Application.version+Application.date)+2)

try:
# Try to open configFile
    with open(configFile, "r"): pass
    
except IOError:
# Except if configFile does not exist, create an example configFile to work from
    try:
        with open (configFile, "w") as exampleFile:
            print
            print "--> Config file not found; Creating "+configFile+"."
            exampleFile.write("## BuildVRFIndex.py CONFIGURATION FILE ##\n#\n")
            exampleFile.write("[account]\n#password is base64 encoded! Plain text passwords WILL NOT WORK!\n#Use website such as http://www.base64encode.org/ to encode your password\nusername=\npassword=\n#\n")
            exampleFile.write("[BuildVRFIndex]\n#variable=C:\path\\to\\filename.ext\nrouterFile=routers.txt\nindexFile=index.txt\nindexFileTmp=index.txt.tmp\nlogFileDirectory=\n")

    # Exception: file could not be created
    except IOError:
        print
        print "--> An error occurred creating the example "+configFile+"."

finally:
# Finally, using the provided configFile (or example created), pull values
# from the config and login to the router(s)
    config = ConfigParser(allow_no_value=True)
    config.read(configFile)
    username = config.get("account", "username")
    password = config.get("account", "password")
    routerFile = config.get("BuildVRFIndex", "routerFile")
    indexFile = config.get("BuildVRFIndex", "indexFile")
    indexFileTmp = config.get("BuildVRFIndex", "indexFileTmp")
    logFileDirectory = config.get("BuildVRFIndex", "logFileDirectory")

    # If logFileDirectory is blank, use current working directory
    if logFileDirectory == "":  logFileDirectory = getcwd()

    # If logFileDirectory does not contain trailing backslash, append one
    if logFileDirectory != "":
        if logFileDirectory[-1:] != "\\": logFileDirectory = logFileDirectory+"\\"
    
    # Define "date" variable for use in the output filename
    date = datetime.now()    # Determine today's date
    date = date.strftime("%Y%m%d")    # Format date as YYYYMMDD

    if fileExist(routerFile):
    # If the routerFile exists, proceed to login to routers
        if fileExist(indexFile):    # If indexFile exists
            remove(indexFile)       # Remove existing indexFile
        routerLogin()               # Log into router(s)    

    else: # if fileExist(routerFile):
    # Else if routerFile does not exist, create an example file and exit
        try:
            with open (routerFile, "w") as exampleFile:
                print
                print "--> Router file not found; Creating "+routerFile+"."
                print "    Edit this file and restart the application."
                exampleFile.write("## BuildVRFIndex.py ROUTER FILE ##\n#\n")
                exampleFile.write("#Enter a list of hostnames or IP Addresses, one per line.\n#For example:\n")
                exampleFile.write("192.168.1.1\n192.168.1.2\nRouterA\nRouterB\nRouterC\netc...")

        # Exception: file could not be created
        except IOError:
            print
            print "--> Required file "+routerFile+" not found; An error has occurred creating "+routerFile+"."
            print "This file must contain a list, one per line, of Hostnames or IP addresses the"
            print "application will then connect to download the running-config."

print
print "--> Done."
raw_input()    # Pause for user input.